jlm/util/GraphWriter.cpp

```diff
@@ -0,0 +1,1188 @@
+    {
+      out << "label=";
+      PrintIdentifierSafe(out, GetLabel());
+      out << " ";
+    }
+    OutputAttributes(out, AttributeOutputFormat::SpaceSeparatedList);
+    out << "];" << std::endl;
+    break;
+  }
+  default:
+    JLM_UNREACHABLE("Unknown GraphOutputFormat");
+  }
+}
+
+Graph::Graph(GraphWriter & writer)
+    : GraphElement(),
+      Writer_(writer),
+      ParentNode_(nullptr)
+{}
+
+Graph::Graph(GraphWriter & writer, Node & parentNode)
+    : GraphElement(),
+      Writer_(writer),
+      ParentNode_(&parentNode)
+{}
+
+const char *
+Graph::GetIdStub() const
+{
+  return "graph";
+}
+
+Graph &
+Graph::GetGraph()
+{
+  return *this;
+}
+
+GraphWriter &
+Graph::GetGraphWriter()
+{
+  return Writer_;
+}
+
+bool
+Graph::IsSubgraph() const
+{
+  return ParentNode_ != nullptr;
+}
+
+Node &
+Graph::CreateNode()
+{
+  auto * node = new Node(*this);
+  Nodes_.emplace_back(node);
+  return *node;
+}
+
+InOutNode &
+Graph::CreateInOutNode(size_t inputPorts, size_t outputPorts)
+{
+  auto * node = new InOutNode(*this, inputPorts, outputPorts);
+  Nodes_.emplace_back(node);
+  return *node;
+}
+
+ArgumentNode &
+Graph::CreateArgumentNode()
+{
+  auto * node = new ArgumentNode(*this);
+  ArgumentNodes_.emplace_back(node);
+  return *node;
+}
+
+ResultNode &
+Graph::CreateResultNode()
+{
+  auto * node = new ResultNode(*this);
+  ResultNodes_.emplace_back(node);
+  return *node;
+}
+
+Edge &
+Graph::CreateEdge(Port & from, Port & to, bool directed)
+{
+  // Edges must be between ports in this graph
+  JLM_ASSERT(&from.GetGraph() == this);
+  JLM_ASSERT(&to.GetGraph() == this);
+
+  // Edge's constructor informs the ports about the edge
+  auto * edge = new Edge(from, to, directed);
+  Edges_.emplace_back(edge);
+  return *edge;
+}
+
+GraphElement *
+Graph::GetElementFromProgramObject(uintptr_t object)
+{
+  auto it = ProgramObjectMapping_.find(object);
+  if (it != ProgramObjectMapping_.end())
+    return it->second;
+  return nullptr;
+}
+
+void
+Graph::MapProgramObjectToElement(GraphElement & element)
+{
+  JLM_ASSERT(&element.GetGraph() == this);
+
+  uintptr_t object = element.GetProgramObject();
+  JLM_ASSERT(object != 0);
+
+  auto & slot = ProgramObjectMapping_[object];
+  JLM_ASSERT(slot == nullptr && "Trying to map a GraphElement to an already mapped program object");
+  slot = &element;
+}
+
+void
+Graph::RemoveProgramObjectMapping(uintptr_t object)
+{
+  size_t erased = ProgramObjectMapping_.erase(object);
+  JLM_ASSERT(erased == 1);
+}
+
+void
+Graph::Finalize()
+{
+  GraphElement::Finalize();
+
+  for (auto & arg : ArgumentNodes_)
+    arg->Finalize();
+  // Nodes with sub graphs also finalize them
+  for (auto & node : Nodes_)
+    node->Finalize();
+  for (auto & res : ResultNodes_)
+    res->Finalize();
+  for (auto & edge : Edges_)
+    edge->Finalize();
+}
+
+void
+Graph::Output(std::ostream & out, jlm::util::GraphOutputFormat format, size_t indent)
+{
+  JLM_ASSERT(IsFinalized());
+
+  switch (format)
```

Comment:
Could we refactor the code in the cases of the switch statement into individual functions?